config-linux.md

```diff
@@ -115,93 +107,59 @@ There is a limit of 5 mappings which is the Linux kernel hard limit.
 
 ## Devices
 
-`devices` is an array specifying the list of devices to be created in the container.
+`devices` is an array specifying the list of devices that MUST be available in the container.
+The runtime may supply them however it likes (with [mknod][mknod.2], by bind mounting from the runtime mount namespace, etc.).
 
 The following parameters can be specified:
 
-* **`type`** *(char, required)* - type of device: `c`, `b`, `u` or `p`. More info in `man mknod`.
-
-* **`path`** *(string, optional)* - full path to device inside container
-
-* **`major, minor`** *(int64, required)* - major, minor numbers for device. More info in `man mknod`. There is a special value: `-1`, which means `*` for `device` cgroup setup.
-
-* **`permissions`** *(string, optional)* - cgroup permissions for device. A composition of `r` (*read*), `w` (*write*), and `m` (*mknod*).
-
-* **`fileMode`** *(uint32, optional)* - file mode for device file
-
-* **`uid`** *(uint32, optional)* - uid of device owner
-
-* **`gid`** *(uint32, optional)* - gid of device owner
-
-**`fileMode`**, **`uid`** and **`gid`** are required if **`path`** is given and are otherwise not allowed.
+* **`type`** *(char, required)* - type of device: `c`, `b`, `u` or `p`.
+  More info in [mknod(1)][mknod.1].
+* **`path`** *(string, required)* - full path to device inside container.
+* **`major, minor`** *(int64, required)* - [major, minor numbers][devices] for the device.
+* **`fileMode`** *(uint32, required)* - file mode for the device.
+  You can also control access to devices [with cgroups](#device-whitelist).
```

Comment:
What I was saying was that, it is possible to use the struct without
specifying this field. Hence it is optional.
On Wed, Jan 27, 2016 at 1:35 PM Vishnu Kannan <vishnuk@google.com> wrote:

> Defaults can also depend on it setgid bit is set on the parent dir I
> think. But it shouldn't matter from the Spec level.
>
> On Wed, Jan 27, 2016 at 1:31 PM W. Trevor King <notifications@github.com>
> wrote:
>
>> In config-linux.md
>> <https://github.com/opencontainers/specs/pull/298#discussion_r51053200>:
>>
>> > -
>> > -* **`permissions`** *(string, optional)* - cgroup permissions for device. A composition of `r` (*read*), `w` (*write*), and `m` (*mknod*).
>> > -
>> > -* **`fileMode`** *(uint32, optional)* - file mode for device file
>> > -
>> > -* **`uid`** *(uint32, optional)* - uid of device owner
>> > -
>> > -* **`gid`** *(uint32, optional)* - gid of device owner
>> > -
>> > -**`fileMode`**, **`uid`** and **`gid`** are required if **`path`** is given and are otherwise not allowed.
>> > +* **`type`** *(char, required)* - type of device: `c`, `b`, `u` or `p`.
>> > +  More info in [mknod(1)][mknod.1].
>> > +* **`path`** *(string, required)* - full path to device inside container.
>> > +* **`major, minor`** *(int64, required)* - [major, minor numbers][devices] for the device.
>> > +* **`fileMode`** *(uint32, required)* - file mode for the device.
>> > +  You can also control access to devices [with cgroups](#device-whitelist).
>>
>> On Wed, Jan 27, 2016 at 01:13:14PM -0800, Vish Kannan wrote: > +*
>> **`fileMode`** *(uint32, required)* - file mode for the device. > + You can
>> also control access to devices [with cgroups](#device-whitelist). Are the
>> following fields required?
>> I'm fine making fileMode, uid, and gid optional. Would they default to
>> the effective uid/gid/umask of the runtime process? Of the container
>> process as it will be when it execs any user-configured arguments?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/opencontainers/specs/pull/298/files#r51053200>.
>>
>